File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.go
var track = make([]int, 26)

func longestIdealString(s string, k int) int {
/*
Bruteforce approach:
- backtrack, exhaust on all possible combination, to choose s[i] or not.
- this is not feasible as it has 2^n time complexity, where n = len(s)

observations:
- let dp[i] = longest ideal subsequence that ends with s[i]
dp[i]
- at index i, 
*/

/*
- k=2 acfgbd -> ac[f][g]bd
- k=2 dcfg
- k=2 acfghi
*/

    clearArr[int](track, 0)
    res := 1
    for i:=0;i<len(s);i++{
        c := int(s[i]-'a')
        currMax := track[c]
        left, right := max(0, c-k), min(25, c+k)
        for l := left; l<=right;l++{
            if currMax < track[l]{
                currMax = track[l]
            }
            if currMax == i {
                break
            }
        }
        track[c] = currMax + 1
        if res < track[c] {
            res = track[c]
        }
    }
    return res
}

func clearArr[T any](arr []T, val T) {
    for i:=0;i<len(arr);i++{
        arr[i] = val
    }
}

File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.rs

use std::cmp::{max, min};

impl Solution {
    pub fn longest_ideal_string(s: String, k: i32) -> i32 {
        let mut track: [i32; 26] = [0; 26];
        let mut longest: i32 = 0;
        let s = s.as_bytes();

        for (i, &c) in s.iter().enumerate() {
            let current_char = (c - 'a' as u8) as i32;
            let mut prev_max = 0;

            let (left, right) = (
                max(0, current_char - k) as usize,
                min(25, current_char + k) as usize,
            );

            for prev in left..right + 1 {
                if prev_max < track[prev] {
                    prev_max = track[prev];
                }
                if prev_max == i as i32 {
                    break;
                }
            }
            track[current_char as usize] = prev_max + 1;
            if longest < track[current_char as usize] {
                longest = track[current_char as usize];
            }
        }

        return longest;
    }
}
